### PR TITLE
hw1_check_pass

### PR DESCRIPTION
**Pull request by KhrystynaTkachyk**
@kudinov-fedor

### CODE REVIEW

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import pytest
+from tests.khrystynatk.hw1_khr_check_password import check_password
+
+
+@pytest.mark.parametrize("pwd, res", [
+    ("sdfjJLsj123!@#", True),
+    ("abc", False),                                         #check pwd < 8 symbols
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:7]: put space after # : 

# check pwd < 8 symbols

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+@pytest.mark.parametrize("pwd1, res1", [
+    ("sdfjJLsj123!@#sdfjJLsj123!@#sd", True),               #catch error for pwd = 30 symbols exactly
+    ("ssdfjJLsj1.,=", True),                                #catch error for ",.=" not being considered as symbols
+    ("sdfjJLsj4!", True),                                   #catch error for "4" not being considered as number
+    ("jTEST5)=", True),                                     #catch error for 'j' is not listed in lowercase letters
+    ("abcLLL3$", True)                                      #catch error for 'L' is not listed in uppercase letters
+])
+def test_error(pwd1, res1):
+    with pytest.raises(AssertionError):
+        assert check_password(pwd1) == res1
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:34]: you do not need here 2nd param in parametrize since it is always the same:

parametrize for 1 param will look like this:

`@pytest.mark.parametrize("pwd1", [
    "sdfjJLsj123!@#sdfjJLsj123!@#sd",               #catch error for pwd = 30 symbols exactly
    "ssdfjJLsj1.,=",                             #catch error for ",.=" not being considered as symbols
    "sdfjJLsj4!",                                   #catch error for "4" not being considered as number
    "jTEST5)=",                                     #catch error for 'j' is not listed in lowercase letters
    "abcLLL3$"                                      #catch error for 'L' is not listed in uppercase letters
])
def test_error(pwd1):
    with pytest.raises(AssertionError):
        assert check_password(pwd1) == True
        # OR
        assert check_password(pwd1)

Comment (kudinov-fedor) [reply tests/khrystynatk/test_hw1_khr_check_password.py:34]: but here error is risen not by function but by you in test
you can try to pass INT into function and check if it works or fails

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+def test_check_pass(pwd, res):
+    assert check_password(pwd) == res
+
+
+@pytest.mark.parametrize("pwd1, res1", [
+    ("sdfjJLsj123!@#sdfjJLsj123!@#sd", True),               #catch error for pwd = 30 symbols exactly
+    ("ssdfjJLsj1.,=", True),                                #catch error for ",.=" not being considered as symbols
+    ("sdfjJLsj4!", True),                                   #catch error for "4" not being considered as number
+    ("jTEST5)=", True),                                     #catch error for 'j' is not listed in lowercase letters
+    ("abcLLL3$", True)                                      #catch error for 'L' is not listed in uppercase letters
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:30]: but what if error with other letter, can you suggest exhaustive test which will cover all letters

File: tests/khrystynatk/hw1_khr_check_password.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+                has_big_letters,
+                has_numbers,
+                has_symbols,
+                not_short,
+                not_long])
+
+
+assert check_password("sdfjJLsj123!@#") is True
+
+print("All checks passed")
```
Comment (kudinov-fedor) [inline tests/khrystynatk/hw1_khr_check_password.py:33]: remove lines 31, 33

File: tests/khrystynatk/test_hw1_khr_check_password.py (hunk 1 above)
Comment (kudinov-fedor) [reply tests/khrystynatk/test_hw1_khr_check_password.py:7]: "abc"     is not a boundary value, and it is not pure length test, it can fail due to missing letter or spec symbol

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+@pytest.mark.parametrize("pwd, res", [
+    ("sdfjJLsj123!@#", True),
+    ("abc", False),                                         #check pwd < 8 symbols
+    ("12345678", False),                                    #check numbers only
+    ("123_5678", False),                                    #check numbers and symbols
+    ("TestPassword", False),                                #check letters only
+    ("TestP@$$word", False),                                #check letterrs and symbols
+    ("TestP2ssw0rd", False),                                #check letters and numbers
+    ("!@#$%^&*()", False),                                  #check symbols only
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:13]: try reverse logic :
- check password without letters
- check password without numbers 
- check ...

test 1 thing at a time, tests will be more clear

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 6]
```diff
@@ -3,32 +3,46 @@
...
+    ("123_5678", False),  # check pwd has no letters
+    ("TestPassword", False),  # check letters only
+    ("TestP@$$word", False),  # check pwd has no numbers
+    ("TestP2ssw0rd", False),  # check pwd has no symbols
+    ("!@#$%^&*()", False),  # check symbols only
+    ("aA!1", False),  # check pwd < 8 symbols
+    ("AAAaaa!1", True),  # check pwd = 8 symbols
+    ("AAAaaa!1" + "1" * 23, False),  # check pwd > 30 symbols
+    ("", False),  # check empty str
+    ("   A1^a   ", True)  # check spaces are counted as chars
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:17]: I would probably expect False in such case,  i.e. ignore spaces at start and end, so that password is only 4 chars long

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 7]
```diff
@@ -3,32 +3,46 @@
...
-def test_error(pwd1, res1):
-    with pytest.raises(AssertionError):
-        assert check_password(pwd1) == res1
+def test_lower_upper_case(func, pwd1, data1, res1):    # check pwd has lower/uppercase letters
+    passwd = pwd1
+    alphabet = data1
+    expected = res1
+    for char in passwd:
+        if char in alphabet:
+            expected = res1
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:46]: expected is always res1

**Review by kudinov-fedor — CHANGES_REQUESTED** [review]
hard to understand what you try to achieve with test_lower_upper_case test

File: tests/khrystynatk/hw1_khr_check_password.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    not_short = len(word) >= 8
+    not_long = len(word) < 30
+
+    return all([has_letters,
+                has_big_letters,
+                has_numbers,
+                has_symbols,
+                not_short,
+                not_long])
+#
```
Comment (kudinov-fedor) [inline tests/khrystynatk/hw1_khr_check_password.py:29]: there should be empty line at end
try to use flake8 it can advise changes required to match expected codestyle

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    ("AAAaaa!1", True),  # check pwd = 8 symbols
+    ("AAAaaa!1" + "1" * 23, False),  # check pwd > 30 symbols
+    ("", False),  # check empty str
+    ("   A1^a   ", True)  # check spaces are counted as chars
+])
+def test_check_pass(pwd, res):
+    assert check_password(pwd) == res
+
+
+@pytest.mark.skip(reason="not implemented")
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:23]: is it skipped just to check skip mark ?

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    with pytest.raises(TypeError):
+        check_password(12345678)
+
+
+@pytest.mark.parametrize("func", [check_password])
+@pytest.mark.parametrize("pwd1, data1, res1", [
+    ("AAABBB1@", "abcdefghijklmnopqrstuvwxyz", False),
+    ("AAaAAA1!", "abcdefghijklmnopqrstuvwxyz", True),
+    ("AAaAAA1!", "ABCDEFGHIJKLMNOPQRSTUVWXYZ", True),
+    ("aaa123$%", "ABCDEFGHIJKLMNOPQRSTUVWXYZ", False)
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:38]: these cases can actually be n=moved to test_check_pass test

**Comment by KhrystynaTkachyk** [comment]
@kudinov-fedor commited new changes

File: tests/khrystynatk/test_hw1_khr_check_password.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+@pytest.mark.parametrize("pwd1, data1", [
+    ("AAA111!@", "abcdefghijklmnopqrstuvwxyz"),
+    ("aaa111!@", "ABCDEFGHIGKLMNOPQRSTUVWXYZ"),
+    ("AAAaaa#!", "0123456789"),
+    ("AAAaaa12", "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~")
+])
+def test_all_chars(pwd1, data1):
+    for j, v in enumerate(data1):
+        res1 = pwd1 + v
+        assert check_password(res1)
```
Comment (kudinov-fedor) [inline tests/khrystynatk/test_hw1_khr_check_password.py:43]: you can try gather all errors , in your case test will fail on first check
though it will do the job, but you will not see the whole picture

**Review by kudinov-fedor — APPROVED** [review]
